v1/tarball/image.go

```diff
@@ -15,17 +15,203 @@
 package tarball
 
 import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
 	"fmt"
+	"io"
+	"os"
+
+	"github.com/google/go-containerregistry/v1/types"
 
 	"github.com/google/go-containerregistry/v1"
 )
 
-// TODO(mattmoor): This.
-// // image accesses an image from a remote registry
-// type image struct{}
-// var _ v1.Image = (*image)(nil)
+type tarBlob struct {
+	name string
+}
+
+type image struct {
+	path     string
+	manifest *v1.Manifest
+	config   *v1.ConfigFile
+}
+
+var _ v1.Image = (*image)(nil)
 
 // Image exposes an image from the tarball at the provided path.
 func Image(p string) (v1.Image, error) {
-	return nil, fmt.Errorf("NYI: tarball.Image(%v)", p)
+	img := image{
+		path: p,
+	}
+	return &img, nil
+}
+
+func (i *image) FSLayers() ([]v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) DiffIDs() ([]v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) ConfigName() (v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) BlobSet() (map[v1.Hash]struct{}, error) {
+	panic("not implemented")
+}
+
+func (i *image) Digest() (v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) MediaType() (types.MediaType, error) {
+	panic("not implemented")
+}
+
+// Manifest returns the v1.Manifest for the specified image.
+// This method memoizes the result, avoiding repeated reads from the tarball.
+func (i *image) Manifest() (*v1.Manifest, error) {
+	if i.manifest == nil {
+		if err := i.loadManifest(); err != nil {
+			return nil, err
+		}
+	}
+	return i.manifest, nil
+}
+
+// singleManifest is the struct used inside the `manifest.json` file of a `docker save` tarball.
+type singleManifest struct {
+	Config   string
+	RepoTags []string
+	Layers   []string
+}
+
+// tarManifest is the struct used to represent a single image inside a `docker save` tarball.
+type tarManifest []singleManifest
+
+func (i *image) loadManifest() error {
+	mfstBytes, err := extractFileFromTar(i.path, "manifest.json")
+	if err != nil {
+		return err
+	}
+
+	tarMfst := tarManifest{}
+	if err := json.Unmarshal(mfstBytes, &tarMfst); err != nil {
+		return err
+	}
+	if len(tarMfst) != 1 {
```

Comment:
I believe the python client allows passing the tag of the image to load.  Fine to TODO but I think we will want this for working with bundles.

Reply:
Hmm thinking more on this, I wonder if the best way to handle this would be:
1. Creat a ~Bundle type that loads the manifest and lazily accesses images.
1. Image loads that and then either extracts the Singleton or a specifically named image.

WDYT?

I see bundles (tarball) and lists (remote) as weakly supported by the python library (they were after thoughts), and I think this gets us the strong tarball support, while starting to improve the bundle story.